RAJA implementation for sparse matrix

nychiang opened the pull request:
This PR is ready for review.
Note that I will change the base from `fr-mds-dev` to develop once we merge #239 

In this PR, I addressed the following functions and their corresponding unit tests:
1. `hiopMatrixRajaSparseTriplet::copyRowsFrom` 
2. `hiopMatrixRajaSparseTriplet::copy_rows_block_from` 
3. `hiopMatrixRajaSparseTriplet::copySubDiagonalFrom`
4. `hiopMatrixRajaSparseTriplet::setSubDiagonalTo`
5. `hiopVectorIntRaja::copy_from`

According to the discussion with Cosmin, I will address other RAJA functions in different PRs.
Try to make each PR as short as possible, so others can review the code easily.

CLOSE #103

pelesh commented:
A few compile warnings that would be good to fix:
```shell
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(342): warning: variable "nx" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(351): warning: variable "M1nnz" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(352): warning: variable "M2nnz" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(677): warning: variable "nrows" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(678): warning: variable "ncols" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(842): warning: variable "nx" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(990): warning: variable "itnz_src" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(991): warning: variable "itnz_dst" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(1076): warning: variable "itnz_src" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(1077): warning: variable "itnz_dst" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(1258): warning: variable "irow_c" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(1260): warning: variable "irow_d" was declared but never referenced
hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp(1274): warning: variable "nnz_Jac_d_new" was declared but never referenced
```

pelesh commented:
Another compile warning for RAJA example 4:
```shell
hiop/src/Drivers/nlpMDS_raja_ex4.hpp(46): warning: overloaded virtual function "hiop::hiopInterfaceBase::get_starting_point" is only partially overridden in class "Ex4"
```

pelesh commented:
Also some warnings in NLP formulation class declaration:
```shell
hiop/src/Optimization/hiopNlpFormulation.hpp(460): warning: type qualifier on return type is meaningless
hiop/src/Optimization/hiopNlpFormulation.hpp(461): warning: type qualifier on return type is meaningless
hiop/src/Optimization/hiopNlpFormulation.hpp(462): warning: type qualifier on return type is meaningless
hiop/src/Optimization/hiopNlpFormulation.hpp(463): warning: type qualifier on return type is meaningless
```

pelesh commented:
Example `NlpMixedDenseSparseRaja4_2` segfaults. When running `nlpMDS_ex4_raja.exe` with options `400 100 1` (one-call constraints), the code segfaults in function `copyRowsFrom` at line
```c++
RAJA::inclusive_scan_inplace<hiop_raja_exec>(dst_row_st_init,dst_row_st_init+n_rows+1,RAJA::operators::plus<int>());
```
Full backtrace is
```shell
(gdb) bt
#0  0x00007fffe7c5faf0 in __GI_raise (sig=<optimized out>) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
#1  0x00007fffe7c61e6c in __GI_abort () at abort.c:90
#2  0x00007fffe7ff0644 in __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007fffe7feb364 in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007fffe7feb420 in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00007fffe7feb8e0 in __cxxabiv1::__cxa_throw (obj=<optimized out>, tinfo=0x7fffe8157548 <typeinfo for std::runtime_error>, 
    dest=0x7fffe800c570 <std::runtime_error::~runtime_error()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:93
#6  0x00007ffff7d9c274 in RAJA_ABORT_OR_THROW (str=0x7ffff7e8e7b0 "CUDAassert")
    at exasgd/src/spack/var/spack/environments/hiop-v0-4-2-deps-newell/.spack-env/view/include/RAJA/util/macros.hpp:152
#7  RAJA::cudaAssert (code=<optimized out>, 
    file=0x7ffff7e92150 "exasgd/src/spack/var/spack/environments/hiop-v0-4-2-deps-newell/.spack-env/view/include/RAJA/policy/cuda/scan.hpp", 
    line=75, abort=true)
    at exasgd/src/spack/var/spack/environments/hiop-v0-4-2-deps-newell/.spack-env/view/include/RAJA/policy/cuda/raja_cudaerrchk.hpp:59
#8  0x00007ffff7db3004 in cudaAssert (abort=<optimized out>, line=<optimized out>, file=<optimized out>, code=<optimized out>)
    at exasgd/src/spack/opt/spack/linux-rhel7-power9le/gcc-7.4.0/raja-0.13.0-uj7zop4pvtllxl7bshthvzzb6jsffi2p/include/cub/util_device.cuh:151
#9  RAJA::impl::scan::inclusive_inplace<128ul, false, int*, RAJA::operators::plus<int, int, int> > (begin=0x7fff601cac00, end=<optimized out>, 
    binary_op=...) at exasgd/src/spack/var/spack/environments/hiop-v0-4-2-deps-newell/.spack-env/view/include/RAJA/policy/cuda/scan.hpp:69
#10 0x00007ffff7da32ac in inclusive_scan_inplace<RAJA::policy::cuda::cuda_exec<128>, int*, RAJA::operators::plus<int, int, int> > (end=<optimized out>, 
    begin=0x7fff601cac00, p=..., binop=...)
    at exasgd/src/spack/var/spack/environments/hiop-v0-4-2-deps-newell/.spack-env/view/include/RAJA/pattern/scan.hpp:66
#11 inclusive_scan_inplace<RAJA::policy::cuda::cuda_exec<128>, int*&, int*, RAJA::operators::plus<int, int, int> > ()
    at exasgd/src/spack/var/spack/environments/hiop-v0-4-2-deps-newell/.spack-env/view/include/RAJA/pattern/scan.hpp:377
#12 hiop::hiopMatrixRajaSparseTriplet::copyRowsFrom (this=0x110ab960, src_gen=..., rows_idxs=0x7fff6006ea00, n_rows=<optimized out>)
    at src/hiop/hiop/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1028
#13 0x00007ffff7dbaf28 in copyRowsFrom (n_rows=400, rows_idxs=0x7fff6006ea00, src_in=..., this=0x110ab940)
    at src/hiop/hiop/src/LinAlg/hiopMatrixMDS.hpp:64
#14 hiop::hiopNlpMDS::eval_Jac_c_d_interface_impl (this=0x7fffffffc930, x=..., new_x=<optimized out>, Jac_c=..., Jac_d=...)
    at src/hiop/hiop/src/Optimization/hiopNlpFormulation.cpp:1352
#15 0x00007ffff7db5680 in hiop::hiopNlpFormulation::eval_Jac_c_d (this=0x7fffffffc930, x=..., new_x=<optimized out>, Jac_c=..., Jac_d=...)
    at src/hiop/hiop/src/Optimization/hiopNlpFormulation.cpp:814
#16 0x00007ffff7dc9d3c in hiop::hiopAlgFilterIPMBase::evalNlp_noHess (this=0x7fffffffcc90, iter=..., f=<optimized out>, c=..., d=..., gradf=..., Jac_c=..., 
    Jac_d=...) at src/hiop/hiop/src/Optimization/hiopAlgFilterIPM.cpp:548
#17 0x00007ffff7dca140 in hiop::hiopAlgFilterIPMBase::startingProcedure (this=0x7fffffffcc90, it_ini=..., f=@0x7fffffffcd60: 398.00004901999984, c=..., 
    d=..., gradf=..., Jac_c=..., Jac_d=...) at src/hiop/hiop/src/Optimization/hiopAlgFilterIPM.cpp:380
#18 0x00007ffff7dcf124 in hiop::hiopAlgFilterIPMNewton::run (this=0x7fffffffcc90)
    at src/hiop/hiop/src/Optimization/hiopAlgFilterIPM.cpp:1348
#19 0x0000000010005c5c in main (argc=<optimized out>, argv=<optimized out>) at src/hiop/hiop/src/Drivers/nlpMDS_ex4_raja_driver.cpp:153
```

nychiang commented:
@pelesh Can you show me how to reproduce the `NlpMixedDenseSparseRaja4_2` segfault? I see that it passes all the pipeline tests on newell [(here)](https://gitlab.pnnl.gov/exasgd/solvers/hiop-mirror/-/pipelines/7902), and it also passes all the unit tests on ascent.

nychiang commented on src/LinAlg/hiopVectorIntRaja.cpp:
```diff
@@ -106,4 +106,13 @@ void hiopVectorIntRaja::copy_to_dev()
   }
 }
 
+void hiopVectorIntRaja::copy_from(const index_type* v_local)
```
I can create a copy constructor in classes "hiopVectorIntSeq" and "hiopVectorIntRaja", but probably we still can keep this function since it can update the vector value.
BTW, I changed some other implementations and hence right now this `copy_from` function is not used in HiOp.

cnpetra replied:
I think copy constructor would not replace the functionality of the method entirely.  But it would be nice to have it go and do not allow to copy to or from raw pointers

ashermancinelli replied:
Could we document the difference? I'm not sure I can tell from the header what the difference is. Perhaps if we don't want a copy constructor we could delete it so we never misuse it.

nychiang replied:
I think Cosmin agree to have copy constructors in the corresponding classes and we can open another PR to add them.

We may need a function to copy from a raw pointer, since we have `iterate_callback` function which allow user to change the values of `hiopVector` between two iterations. See `hiopVector::copyFrom(double*)` for an example.
Therefore, I follow the same pattern to create this function, i.e., `hiopVectorInt::copy_from(index_type*)`.
This function was used in unit test,  but now I am using new implementations and this one is not used anymore.

nychiang replied:
I checked the code again. Please let me correct my reply above.
`hiopVector::copyFrom(double*)` is NOT used in the `iterate_callback`, but used in some callback functions in Ex1 and primal-decomposition solver. It is also used in some MPI functions. For example, copy back the MPI_All_Reduce results to a vector.

Again, I am fine to delete this function for this PR right now since it is not used. Or just leave it for now and create another issue to address it by a copy constructor. It is not urgent at all, since we may not need this function in the near future.

cnpetra replied:
yes, fill an issue for copy constructor (my opinion is that we should not support them). Then let's ensure `copy_from` works fine and try to merge this.

pelesh replied:
@cnpetra, isn't `new_copy` method in `hiopVector` supposed to create a deep copy of `*this`? Perhaps having a similar method/implementation in the `hioVectorInt` would be a consistent solution?

cnpetra replied:
yes, `new_copy` and `alloc_clone` are preferred to having copy constructors...

nychiang commented:
@pelesh Some variables shown in the compiler warning are used in `assert` function. For example, https://github.com/LLNL/hiop/blob/69db8d477c04083eb2381ac058d107c16f623223/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L342-L345 and

nychiang commented:
@pelesh  I will rebase this branch to develop, after PR #263. Then I will run the PNNL CI, to see if it can reproduce the segfault you mentioned.

nychiang commented:
@pelesh I will rebase this branch to `issue267-fix`, otherwise merging these two branches to `develop` will introduce some conflicts.

nychiang commented:
@pelesh it seems that branch `issue267-fix` is still under development. Hence I rebase my branches to `develop`.
I removed all the changes I added to remove the compiler warnings, to reduce the conflicts when we merge your PR #268.
In addition, I cannot reproduce the segfault you mentioned above on ascent. Can you try the latest code again to see if the problem still exists?

nychiang commented:
@pelesh 
I can reproduce the runtime error on line
`RAJA::inclusive_scan_inplace<hiop_raja_exec>(dst_row_st_init,dst_row_st_init+n_rows+1,RAJA::operators::plus<int>());`
This failure comes from `CUDAassert`, and it only happens on mariana. (All the unit tests work fine on Newell.)

Seems like I am using CPU code again in this line, but I really have no idea where I made the mistake.

@ashermancinelli or @pelesh 
Could you help me to check why this branch fails again on mariana? Any detailed log file may help.

pelesh commented:
> @ashermancinelli or @pelesh
> Could you help me to check why this branch fails again on mariana? Any detailed log file may help.

I am looking into it ...

nychiang commented:
@pelesh @ashermancinelli @cnpetra 
I tried to narrow down the problem, but haven't fixed it yet.
In short, now the CUDAassert error can be reproduced in the unit test, by setting `mem_space` to `um` in the [source code of unit test](https://github.com/LLNL/hiop/blob/sparse-gpu-kernel-dev/tests/testMatrixSymSparse.cpp#L223):
`options.SetStringValue("mem_space", "device");` 
-->
`options.SetStringValue("mem_space", "um");`

The error code I obtained from a gdb session is `cudaErrorInvalidDevice = 101`.
[This indicates that the device ordinal supplied by the user does not correspond to a valid CUDA device or that the action requested is invalid for the specified device.](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html)

Since setting `mem_space` to `device` passes all the unit tests, I am guessing this issue is related to how RAJA/CUDA is configured on the platform. Any thoughts?

cnpetra commented:
> @pelesh @ashermancinelli @cnpetra
> I tried to narrow down the problem, but haven't fixed it yet.
> In short, now the CUDAassert error can be reproduced in the unit test, by setting `mem_space` to `um` in the [source code of unit test](https://github.com/LLNL/hiop/blob/sparse-gpu-kernel-dev/tests/testMatrixSymSparse.cpp#L223):
> `options.SetStringValue("mem_space", "device");`
> -->
> `options.SetStringValue("mem_space", "um");`
> 
> The error code I obtained from a gdb session is `cudaErrorInvalidDevice = 101`.
> [This indicates that the device ordinal supplied by the user does not correspond to a valid CUDA device or that the action requested is invalid for the specified device.](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html)
> 
> Since setting `mem_space` to `device` passes all the unit tests, I am guessing this issue is related to how RAJA/CUDA is configured on the platform. Any thoughts?

that's weird. We should isolate the issue even more, like in a standalone driver. Where do you get the cuda assert exactly?

nychiang commented:
@cnpetra 
At function [`RAJA::inclusive_scan_inplace` ]("https://github.com/LLNL/hiop/blob/e2afa72704fcd27db2f03a1887c3f27d9d7901f2/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L1019"). However, I just noticed that the latest code passed all the PNNL pipeline tests. 
@ashermancinelli Did we changed the module settings on mariana? Same code failed on mariana but now it is fine.

cnpetra commented:
> @cnpetra
> At function [`RAJA::inclusive_scan_inplace`](https://github.com/LLNL/hiop/blob/e2afa72704fcd27db2f03a1887c3f27d9d7901f2/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L1019). However, I just noticed that the latest code passed all the PNNL pipeline tests.
> @ashermancinelli I did rebase this branch onto develop. Did we changed the module settings on mariana? Same code previously failed on mariana but now it is fine.

yeap, difficult to reproduce. even on lassen, fails only on the compute note. I pm'ed you about this.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.hpp:
```diff
@@ -143,75 +173,82 @@ class hiopMatrixRajaSparseTriplet : public hiopMatrixSparse
-  {
-    assert(false && "not needed / implemented");
-  }
+                                 const bool offdiag_only = false) {assert(0 && "FIXME_NY");}
```
fix assert message and code styling.

nychiang replied:
These functions require RAJA implementations, which are filled in other PRs.
I changed the description in assert, to highlight these functions and hence I can quickly search for them.
For example, this function `copySubmatrixFrom` is implemented in `sparse-gpu-kernel-2-dev`, and you can [see this assert is removed from the code](https://github.com/LLNL/hiop/blob/6cf5c1087378c9b478cb38ced32f569ace399791/src/LinAlg/hiopMatrixRajaSparseTriplet.hpp#L202-L206).

cnpetra replied:
ok, I see. But the ones below are not removed in `sparse-gpu-kernel-2-dev. My point was to use a more suggestive assert message "not yet implemented".

nychiang replied:
They are removed in sparse-gpu-kernel-3-dev and sparse-gpu-kernel-4-dev. :-)
I will change the message to "not yet implemented" and rebase the following branches.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.hpp:
```diff
@@ -143,75 +173,82 @@ class hiopMatrixRajaSparseTriplet : public hiopMatrixSparse
-  {
-    assert(false && "not needed / implemented");
-  }
+                                                       const hiopVector& ix) {assert(0 && "FIXME_NY");}
```
fix assert message and code styling.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.hpp:
```diff
@@ -143,75 +173,82 @@ class hiopMatrixRajaSparseTriplet : public hiopMatrixSparse
-  {
-    assert(false && "not needed / implemented");
-  }
+                                                       const hiopVector& ix) {assert(0 && "FIXME_NY");}
```
fix assert message and code styling.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.hpp:
```diff
@@ -143,75 +173,82 @@ class hiopMatrixRajaSparseTriplet : public hiopMatrixSparse
-  {
-    assert(false && "not needed / implemented");
-  }
+                                        const int &nnz_to_copy){assert(0 && "FIXME_NY");}
```
fix assert message and code styling.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.hpp:
```diff
@@ -143,75 +173,82 @@ class hiopMatrixRajaSparseTriplet : public hiopMatrixSparse
-  {
-    assert(false && "not needed / implemented");
-  }
+                                                  const hiopVector& pattern) {assert(0 && "FIXME_NY");}
```
fix assert message and code styling.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.hpp:
```diff
@@ -238,10 +275,7 @@ class hiopMatrixRajaSparseTriplet : public hiopMatrixSparse
-  {
-    assert(false && "not needed / implemented");
-  }
+                            int **index_covert_extra_Diag2CSR) {assert(0 && "FIXME_NY");}
```
fix assert message and code styling.

cnpetra commented:
> > @cnpetra
> > At function [`RAJA::inclusive_scan_inplace`](https://github.com/LLNL/hiop/blob/e2afa72704fcd27db2f03a1887c3f27d9d7901f2/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L1019). However, I just noticed that the latest code passed all the PNNL pipeline tests.
> > @ashermancinelli I did rebase this branch onto develop. Did we changed the module settings on mariana? Same code previously failed on mariana but now it is fine.
> 
> yeap, difficult to reproduce. even on lassen, fails only on the compute note. I pm'ed you about this.

It appears that upgrading raja to 0.13 and umpire to 5.0 fixes this problem on lassen as well. For the record, the scan fails with raja 0.13 and umpire 4.0 as well as raja 0.10 and umpire 4.0.

nychiang commented:
Reopen

cnpetra commented:
not sure how to reopen it
